File: packages/nativescript-smartlook/platforms/ios/src/NSCSmartlook.swift
import Foundation
import UIKit
import SmartlookAnalytics

@objc public class NSCSmartlook: NSObject {
    @objc public static func start(_ key: String) {
        Smartlook.instance.preferences.projectKey = key
        Smartlook.instance.start()
    }

    @objc public static func stop() {
        Smartlook.instance.stop()
    }

    @objc public static func isRecording() -> Bool {
        return Smartlook.instance.state.status == .recording
    }

    @objc public static func setUser(id: String, name: String?, email: String?, extraData: NSDictionary?) {
        Smartlook.instance.user.identifier = id
        if (name != nil) {
            Smartlook.instance.user.name = name
        }
        if (email != nil) {
            Smartlook.instance.user.email = email
        }
        if (extraData != nil) {
            for (key, value) in extraData! {
                Smartlook.instance.user.setProperty(key as! String, to: value as? String)
            }

        }
    }

    @objc public static func sessionUrl(_ withTimestamp: Bool) -> String {
        if (withTimestamp == true) {
            return Smartlook.instance.user.session.urlWithTimestamp!.absoluteString
        } else {
            return Smartlook.instance.user.session.url!.absoluteString
        }
    }

    @objc public static func setRenderingMode(_ mode: NSNumber) {
        switch mode.intValue {
        case 0:
            Smartlook.instance.preferences.renderingMode = .native
            break;
            case 1:
            Smartlook.instance.preferences.renderingMode = .wireframe()
            break;
            case 2:
            Smartlook.instance.preferences.renderingMode = .noRendering
            break;
        default:
            Smartlook.instance.preferences.renderingMode = .native
            break;

        }
    }

    @objc public static func getRenderingMode() -> NSNumber {
        switch(Smartlook.instance.preferences.renderingMode) {
        case .native:
            return 0;
        case .wireframe:
            return 1;
        case .noRendering:
            return 2;
        default:
            return 0;
        }
    }

    @objc public static func setSensitivity(_ view: UIView, sensitive: Bool) {
        view.slSensitive = sensitive
    }

    @objc public static func trackEvent(_ name: String, properties: NSDictionary?) {
        if (properties != nil) {
            let props = Properties()
            for (key, value) in properties! {
                props.setProperty(key as! String, to: value as? String)
            }
            Smartlook.instance.track(event: name, properties: props)
        } else {
            Smartlook.instance.track(event: name)
        }
    }

    @objc public static func trackNavigationEvent(_ name: String, properties: NSDictionary?) {
        if (properties != nil) {
            let props = Properties()
            for (key, value) in properties! {
                props.setProperty(key as! String, to: value as? String)
            }
            Smartlook.instance.track(navigationEvent: name, properties: props)
        } else {
            Smartlook.instance.track(navigationEvent: name)
        }
    }
}

File: packages/nativescript-smartlook/platforms/android/java/io/nstudio/plugins/smartlook/NSCSmartlook.kt
package io.nstudio.plugins.smartlook

import android.view.View
import com.smartlook.android.core.api.Smartlook
import com.smartlook.android.core.api.extension.isSensitive
import com.smartlook.android.core.video.annotation.RenderingMode
import com.smartlook.android.core.api.model.Properties

class NSCSmartlook {

    companion object {
        @JvmStatic
        fun start(key: String) {
            Smartlook.instance.preferences.projectKey = key
            Smartlook.instance.start()
        }

        @JvmStatic
        fun stop() {
            Smartlook.instance.stop()
        }

        @JvmStatic
        fun isRecording(): Boolean {
            return Smartlook.instance.state.status.isRecording
        }

        @JvmStatic
        fun sessionUrl(withTimestamp: Boolean): String {
          return if (withTimestamp) {
            Smartlook.instance.user.session.urlWithTimestamp.toString()
          } else {
            Smartlook.instance.user.session.url.toString()
          }
        }

        @JvmStatic
        fun setRenderingMode(mode: Int) {
            when (mode) {
                0 -> Smartlook.instance.preferences.renderingMode = RenderingMode.NATIVE
                1 -> Smartlook.instance.preferences.renderingMode = RenderingMode.WIREFRAME
                2 -> Smartlook.instance.preferences.renderingMode = RenderingMode.NO_RENDERING
            }
        }


/*
// todo
        @JvmStatic
        fun getRenderingMode(): Int {
            return when (Smartlook.instance.state.renderingMode) {
                RenderingMode.NATIVE -> 0
                RenderingMode.WIREFRAME -> 1
                RenderingMode.NO_RENDERING -> 2
            }
        }
*/


        @JvmStatic
        fun setSensitivity(view: View, sensitive: Boolean) {
            view.isSensitive = sensitive
        }

        @JvmStatic
        fun setUser(id: String, name: String?, email: String?, extraData: HashMap<String, String>?) {
            Smartlook.instance.user.identifier = id
            if (name != null) {
                Smartlook.instance.user.name = name
            }
            if (email != null) {
                Smartlook.instance.user.email = email
            }
            if (extraData != null) {
                for (prop in extraData) {
                    Smartlook.instance.user.properties.putString(prop.key, prop.value)
                }
            }
        }

        @JvmStatic
        fun trackEventProperties(eventName: String, mapProperties: HashMap<String, String>?) {
            val properties = propertiesFromMap(mapProperties)

            Smartlook.instance.trackEvent(eventName, properties)
        }

        @JvmStatic
        fun trackNavigationEvent(screenName: String, mapProperties: HashMap<String, String>?) {
            val properties = propertiesFromMap(mapProperties)

            Smartlook.instance.trackNavigationEnter(screenName, properties)
        }

        private fun propertiesFromMap(propertiesMap: HashMap<String, String>?): Properties? {
            if (propertiesMap == null) {
                return null
            }

            val properties = Properties()

            for (prop in propertiesMap) {
                properties.putString(prop.key, prop.value)
            }

            return properties
        }
    }
}
